### ISSUE TITLE
[批次] Q4 KV：packed int4 存储、paged kernel 与 runtime

### ISSUE THREAD

**Issue by gqf2008** [gqf2008/machserve#192]
## 目标

为 dense、非 MLA 的 full-attention KV 增加 packed int4 存储格式（Q4 KV），先建立 CPU oracle 和 paged 布局，再实现 HIP kernel，最后接 `MACH_KV=q4`。

### Stage 1（本批先做）

- `Q4Kv`：对称 int4，每 `(token, kv_head)` 一个 f32 scale；
- 每 head row 按 2 个 nibble/byte 打包；
- `PagedQ4KvLayout`：page-major `[page, off, kv_head, packed_dim]` payload + `[page, off, kv_head]` scales；
- scatter/gather/quantize/dequantize/dot/attention CPU oracle；
- 覆盖 roundtrip、极值、非有限输入、paged index、GQA attention parity。

### Stage 2

- `kv_store_paged_q4` HIP kernel；
- correctness-first `attn_decode_paged_q4_gqa` HIP kernel；
- 加入 HIPRTC offline 门禁；
- direct kernel CPU-oracle GPU parity（opt-in、fail-loud）。

### Stage 3

- runtime `MACH_KV=q4`，仅 dense 非 MLA、F16 compute；
- 显存预检按 packed payload + scales；
- server/engine 路由与 guard；
- 真机 GPU parity / 性能矩阵。

## 验收

- Q4 payload 大小约为 INT8 的一半（不含 scales）；
- CPU oracle 与 f32/INT8 attention 对比有明确误差上界；
- paged 索引测试覆盖跨页、非整除 head_dim packing（若支持）；
- Stage 2/3 必须有独立 GPU parity，未跑真机前不宣称 GPU 正确。

## 非目标

- MLA Q4 KV；
- 训练/微调；
- 因子分解或其他 attention 近似；
- 直接把现有 INT8 改成 Q4 而不保留独立格式。

## 关联

- #126 / #134 / #142
- #190
- #191

**Comment by gqf2008** [comment]
## Stage 1 已合并

- PR #193 squash merge：`a3c65a3`
- 内容：packed Q4 KV 对称 int4 布局、paged payload/scales、scatter/gather/dequant/dot/attention CPU oracle
- 关键契约：`[-7,7]`，code 8 为 0，low nibble first，每 `(token, kv_head)` 一个 f32 scale；支持 odd `head_dim`
- 覆盖：11 个 Q4 专项测试，含 roundtrip 上界、subnormal scale、paged 跨页/坏页表/超容量、非有限输入、GQA CPU parity
- 独立审查：两轮复核通过，无阻塞问题

## 仍未完成

- Stage 2：`kv_store_paged_q4` / `attn_decode_paged_q4_gqa` HIP kernels + HIPRTC offline 门禁 + opt-in GPU parity
- Stage 3：`MACH_KV=q4` runtime、显存预检、server/engine guard、真机 parity/性能矩阵

本 issue 保持 open，继续推进 Stage 2。

**Comment by gqf2008** [comment]
🔄 处理中 `[wt-paged-kv-q4-stage2]`：开始 Stage 2（`kv_store_paged_q4` + `attn_decode_paged_q4_gqa` + HIPRTC offline + opt-in direct parity）。

**Comment by gqf2008** [comment]
## Stage 2 已合并

- PR #194 squash merge：`58e5e37`
- 新增 `kv_store_paged_q4`、`attn_decode_paged_q4_gqa` 与 Rust launcher；
- HIPRTC offline 门禁 79→81，81 个 kernel 全编译通过；
- 新增 opt-in `q4_kv` GPU parity：store 逐位对拍 CPU、attention 对拍 CPU oracle，覆盖 `dim=65` + 乱序 page table 与 `dim=128` 跨页；
- 独立子代理两轮审查最终 APPROVE；显式 `--ignored` 未设 `MACH_TEST_Q4_KV=1` 时 fail-loud。
- 未跑真机 GPU parity，不宣称 gfx1100 数值已验证。

## 剩余 Stage 3

- `MACH_KV=q4` runtime：dense、非 MLA、F16 compute；
- packed payload + scale 显存预检；
- server/engine 路由与 guard；
- 真机 GPU parity / 性能矩阵。

继续推进 Stage 3（先离线接线，无真机验证不宣称 GPU 正确）。

**Comment by gqf2008** [comment]
🔄 处理中 `[wt-paged-kv-q4-stage3]`：开始 Stage 3（`MACH_KV=q4` runtime、显存预检、server/engine guard；真机 parity 仍待后续 GPU 窗口）。

**Comment by gqf2008** [comment]
## Stage 3 已合并

- PR #195 squash merge：`53fead9`
- `MACH_KV=q4` runtime 已接 paged Q4 store/attention，仅 dense 非 MLA F16 + `MACH_Q4=1 MACH_Q4_DEVICE=2 MACH_PAGED=1`；
- packed payload+scales 显存预检、server/engine fail-fast guard、doctor/runtime 共用 `KvMode` 解析；
- paged Q4 禁用 graph capture；state anchor / 连续 compaction 显式拒绝；
- 双面 check/clippy、workspace tests、HIP 模式专项测试、81-kernel HIPRTC offline 全绿；
- 独立子代理两轮审查最终 APPROVE。

## 仍保持 open

- gfx1100 Q4 KV GPU parity / 性能矩阵；
- 端到端真机 runtime 回归（解析→分配→prefill/decode→page compaction）。
- 未跑真机，不宣称 GPU 数值或性能已验证。

**Comment by gqf2008** [comment]
## 收尾状态（2026-09-15）

离线部分已全部合并到 `master=53fead9`，无待合并 PR、无本线遗留 worktree：

- Stage 1：`#193 / a3c65a3`
- Stage 2：`#194 / 58e5e37`
- Stage 3：`#195 / 53fead9`
- 本地最终门禁：`cargo fmt --all --check`、双面 `cargo check --workspace --all-targets`、`cargo clippy --workspace --lib --tests --features hip -- -D warnings`、`cargo test --workspace` 全绿。
- 主仓唯一未跟踪文件为既有 `crates/mach-model/examples/ab_probe_tmp.rs`，按要求未删除；它会让 `--all-targets` check 产生一个未使用 `mut` warning，但不属于已提交代码。

## 保持 open 的验收项

本 issue 不关闭：验收明确要求 Stage 2/3 有独立真机 GPU parity。当前仍未执行：

- gfx1100 `MACH_KV=q4` store/attention GPU parity；
- 端到端真机 runtime（解析→分配→prefill/decode→page compaction）；
- 性能矩阵。

未来 GPU 窗口入口：

```text
MACH_TEST_Q4_KV=1; cargo test -p mach-model --features hip --test q4_kv -- --ignored --test-threads 1
```

执行前必须遵守既有安全约束：一次只跑一条 GPU arm，结束后查 Display 4101 与 `mach-server doctor` 的 `device_count=2`，留观察窗口再跑下一 arm。未跑真机前不宣称 Q4 KV 的 GPU 数值或性能已验证。

**Comment by gqf2008** [comment]
## 真机 GPU parity（2026-09-15）

硬件：AMD Radeon RX 7900 XTX（gfx1100），`device_count=2`，23.84 GiB free。  
代码：`master=53fead9`。

执行：

```powershell
$env:MACH_TEST_Q4_KV='1'
cargo test -p mach-model --features hip --test q4_kv -- --ignored --test-threads 1 --nocapture
```

结果：

```text
running 2 tests
test paged_q4_attention_matches_cpu_oracle ... ok
test paged_q4_store_matches_cpu_oracle ... ok
test result: ok. 2 passed; 0 failed; 0 ignored
q4_kv_exit=0 elapsed=28.2s
```

覆盖内容：

- `kv_store_paged_q4` 逐位对拍 CPU `Q4Kv::quantize`；
- `attn_decode_paged_q4_gqa` 对拍 CPU `attention_decode_q4`；
- `dim=65` + 乱序 page table；
- `dim=128` 跨页路径。

跑后复检：Display 无新增 4101，Kernel-Power 无新增 41，`mach-server doctor` 仍为 `device_count=2`、gpu[0] 为 RX 7900 XTX。未出现掉卡或整机异常。

结论：Q4 KV direct kernel GPU parity 已实机通过。仍保持 open 的仅是端到端 runtime 回归与性能矩阵；性能未测前不宣称性能。

**Comment by gqf2008** [comment]
## Q4 KV runtime smoke（2026-09-15）

这是 direct kernel parity 之后的第二 arm，定位为 **runtime smoke**，不替代 kernel CPU-oracle parity，也不宣称完整 runtime wiring parity。

命令：

```powershell
$env:MACH_TEST_Q4_KV='1'
cargo test -p mach-model --features hip --test q4_kv batched_model_q4_q4_kv -- --ignored --test-threads 1 --nocapture
```

整理后的运行记录（运行 revision `a3144c1`；当时使用旧测试名 `batched_model_q4_q4_kv_matches_f16_paged_decode` / `batched_model_q4_q4_kv_prefill_matches_f16`；后续 revision 仅重命名/注释，GPU 执行逻辑不变）：

```text
test batched_model_q4_q4_kv_matches_f16_paged_decode ... ok
test batched_model_q4_q4_kv_prefill_matches_f16 ... ok
2 passed; 0 failed; 0 ignored; 0 filtered out
libtest 19.38s; command wall 21.1s
```

覆盖：tiny Q4 dense + paged Q4 KV 的构造、Q4 KV 分配/分派、单 token decode、2-slot mixed prefill；Q4 与 f16 KV logits 做粗粒度有限差 `<1.0` 门禁。精确层数断言是后续 revision 加入的，不包含在这次真机 run 的执行范围；逐位数值正确性由 `paged_q4_store_matches_cpu_oracle` / `paged_q4_attention_matches_cpu_oracle` 承担。

跑后 Display 无新增 4101、Kernel-Power 无新增 41，doctor 仍为 `device_count=2`、gpu[0] 为 RX 7900 XTX。性能矩阵未执行。

**Comment by gqf2008** [comment]
## Runtime smoke 已合并

- PR #197 squash merge：`f6e37b9`
- `batched_model_q4_q4_kv_runtime_smoke_decode` / `..._prefill` 已进入 master（opt-in `MACH_TEST_Q4_KV=1` + `#[ignore]`）。
- 最终复检 2026-09-15 16:58：Display 最新仍为 2026-09-12 4101，Kernel-Power 最新仍为 2026-09-13 41；`doctor` = `device_count=2`，gpu[0] RX 7900 XTX，23.84 GiB free。
- 当前真机覆盖：direct kernel CPU-oracle parity + tiny paged runtime smoke。
- 仍未完成：性能矩阵、真实 checkpoint 端到端 Q4 KV 回归。未跑这些前不宣称性能或完整 runtime parity。

**Comment by gqf2008** [comment]
## 性能矩阵事故：8B paged f16 arm 触发 TDR / 整机异常关机（2026-09-15）

在尝试 Q4 KV 性能矩阵时，先跑 `MACH_KV=f16` 的 Qwen3-8B paged arm：

- Qwen3-8B / Q4-on-device / `MACH_PAGED=1` / `MACH_TPP=64` / capacity 1 / prefill_rows 16 / 约 420-token 共享前缀 + 3 个短 completion；
- server 完成 4 次 admit/register，但因共享前缀没有命中整页复用，脚本按既有门禁拒绝写出性能 JSON；
- 17:24:40 出现 Display 4101（amduw23g 停止响应后恢复），随后系统 Kernel-Power 41 + 6008 异常关机；
- 20:33:52 冷复位后 `mach-server doctor` = `device_count=1`，只剩核显 12.16 GiB，RX 7900 XTX 不在 HIP 枚举中。

结论：当前机器上的 Qwen3-8B paged 性能矩阵不可继续；本轮没有产生可用的 Q4 KV 性能数字。`device_count=1` 期间禁止任何 GPU 负载或重试。

事故已追加到 `LESSON_重GPU真机测试完成后仍可能触发整机崩溃需留观察窗口.md`。Q4 direct kernel parity 与 tiny runtime smoke 仍有效；阻塞的是 8B paged 性能矩阵，不是 kernel numerics。

**Comment by gqf2008** [comment]
## 同后端对照更正：llama.cpp HIP 稳定（2026-09-15）

之前把本机 TDR/掉卡归因于“Vulkan 稳、ROCm Windows 不稳”是不成立的。已在同一台机器、同一 ROCm 6.2 上自行编译并运行 llama.cpp HIP：

- `build-hip`：`GGML_HIP=ON`、`GPU_TARGETS=gfx1100`、ROCm clang 19、hipBLAS；
- 为兼容 ROCm 6.2 只有 `__hip_fp8_e4m3_fnuz`，仅加编译期 `-D__hip_fp8_e4m3=__hip_fp8_e4m3_fnuz`，未改源码；
- `llama-bench -m qwen2.5-0.5b-instruct-q8_0.gguf -p 128 -n 16 -r 1 -ngl 99` 正常完成：
  - backend ROCm，Device 0 = RX 7900 XTX / gfx1100；
  - pp128 4654.8 t/s，tg16 283.0 t/s，3.4s；
  - 跑后无新增 4101/41，`device_count=2`，XTX 仍在。

结论：同一 HIP/ROCm 驱动栈可稳定运行。故障范围应限制在 MachServe 特有路径：hiprtc 运行时内核、paged attention/page table、Q4-on-device GEMV、server 生命周期/teardown。后续用 llama.cpp HIP 作为稳定对照做单变量定位，不再泛化到后端或硬件。

**Comment by gqf2008** [comment]
## llama.cpp HIP 同后端对照：27B Q4_K_M 正常（2026-09-15）

在同一台 7900 XTX / ROCm 6.2 上，使用自行编译的 llama.cpp HIP 后端跑 Qwen3.8-27B Q4_K_M：

- 模型：`qwen3.8-27b-q4_k_m.gguf`，15.40 GiB，26.90B params；
- backend：ROCm，Device 0 = RX 7900 XTX / gfx1100，`ngl=99`；
- 命令：`llama-bench -m qwen3.8-27b-q4_k_m.gguf -p 64 -n 8 -r 1 -ngl 99`；
- 结果：pp64 **537.47 t/s**，tg8 **36.70 t/s**，总耗时 12.2s；
- 跑后无新增 4101/41，XTX 仍在设备枚举中。

此前 MachServe Qwen3-8B paged f16、capacity 1、prefill_rows 16 的 arm 反而触发 4101/异常关机。因此问题不再是“ROCm Windows 或 8B/27B 模型规模不可用”，而应定位到 MachServe 自身路径：

- hiprtc 运行时内核生命周期；
- paged attention/page table；
- Q4-on-device GEMV 与 paged KV 组合；
- mach-server 进程 teardown/同步。

**Comment by gqf2008** [comment]
已迁移到 walgit 线程 `gh-192`:主仓=自建 walgit,GitHub 自 2026-09-19 起仅作镜像与发版,issue 面已关闭。后续进展请看 walgit —— 看板 `walgit --config ~/.walgit/walgit.toml collab board`,线程详情 `walgit collab thread gh-192`。
